crit/mempages.go

```diff
@@ -0,0 +1,134 @@
+package crit
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/checkpoint-restore/go-criu/v6/crit/images/pagemap"
+)
+
+const pageSize = 4096
+
+// MemoryReader is a struct used to retrieve
+// the content of memory associated with a specific process ID (pid).
+// New instances should be created with NewMemoryReader()
+type MemoryReader struct {
+	checkpointDir  string
+	pid            int
+	pagesID        uint32
+	pagemapEntries []*pagemap.PagemapEntry
+}
+
+// NewMemoryReader creates a new MemoryReader instance with all fields populated.
+func NewMemoryReader(checkpointDir string, pid int) (*MemoryReader, error) {
+	pagemapImg, err := getImg(filepath.Join(checkpointDir, fmt.Sprintf("pagemap-%d.img", pid)), &pagemap.PagemapHead{})
+	if err != nil {
+		return nil, err
+	}
+
+	pagesID := pagemapImg.Entries[0].Message.(*pagemap.PagemapHead).GetPagesId()
+
+	pagemapEntries := make([]*pagemap.PagemapEntry, 0)
+
+	for _, entry := range pagemapImg.Entries[1:] {
+		pagemapEntries = append(pagemapEntries, entry.Message.(*pagemap.PagemapEntry))
+	}
+
+	return &MemoryReader{
+		checkpointDir:  checkpointDir,
+		pid:            pid,
+		pagesID:        pagesID,
+		pagemapEntries: pagemapEntries,
+	}, nil
+}
+
+// GetMemPages retrieves content of memory associated with a pid.
+func (mr *MemoryReader) GetMemPages(start, end uint64) (*bytes.Buffer, error) {
+	// Calculate and check the size of the memory area
+	size := end - start
+	if size == 0 {
+		return nil, fmt.Errorf("memory size must be greater than 0")
+	}
+
+	startPage := start / pageSize
+	endPage := end / pageSize
+
+	var buffer bytes.Buffer
```

Comment:
I am not experienced enough in Go, but in C this would be wrong. You are creating a local variable and returning the pointer to it. At the end of the function the variable would be gone. Go has a garbage collector so maybe this is okay.

My expectations from C would be something like:
```go
func (mr *MemoryReader) GetMemPages(buffer *bytes.Buffer, start, end uint64) (error)  {
    // fill buffer
    return nil
}
```

Your approach seems wrong to me, but that could be just that I do not understand Go correctly.

@rst0git any comments from your side

Reply:
The Go compiler does escape analysis during allocation, so it would identify the lifetime of the buffer pointer correctly and allocate it on the heap. You're right, this would be incorrect in a language like C, but is permissible in Go.

Reply:
Thanks for your input. Any suggestion how to implement this. With a local variable like it is or something different?

Reply:
Since both approaches will work, I am not sure which one is idiomatic Go 🤔

Reply:
This pattern is quite common in Go, I've seen it across other projects before. I feel it's fine, the current design should work as intended. The allocator + GC will take care of things.

Reply:
Agree with @snprajwal, I have seen this pattern also in other Go projects.